### PR TITLE
Add Quark Script APIs to detect CWE-532

### PR DESCRIPTION
**Pull request by pulorsok**
# Detect CWE-532 in Android Application (dvba.apk)

This scenario seeks to find **insertion of sensitive information into Log file**. See [CWE-532](https://cwe.mitre.org/data/definitions/532.html) for more details.

Let’s use this [APK](https://github.com/rewanthtammana/Damn-Vulnerable-Bank) and the above APIs to show how the Quark script finds this vulnerability.

First, we use API `findMethodInAPK` to locate the method `log.d`. Then we use API `methodInstance.getArguments` to get the argument that input to `log.d`. Finally, we use keywords such as "token", "password", and "decrypt" to check if arguments include sensitive data. If the answer is YES, that may cause sensitive data leakage into log file.

You can use your own keywords in the keywords list to detect sensitive data.

### findMethodInAPK(samplePath, targetMethod)
- Description: Find the target method in APK
- params: 
    1. samplePath: Target file
    2. targetMethod: A python list contains class name, method name, and descriptor of target method
- return: python list contains caller method instance of target method
### Quark Script CWE-532.py
```python
from quark.script import findMethodInAPK

SAMPLE_PATH = "dvba.apk"
TARGET_METHOD = [
    "Landroid/util/Log;",                       # class name
    "d",                                        # method name
    "(Ljava/lang/String; Ljava/lang/String;)I"  # descriptor
]
CREDENTIAL_KEYWORDS = [
    "token",
    "decrypt",
    "password"
]

methodsFound = findMethodInAPK(SAMPLE_PATH, TARGET_METHOD)

for debugLogger in methodsFound:
    arguments = debugLogger.getArguments()

    for keyword in CREDENTIAL_KEYWORDS:
        if keyword in arguments[1]:
            print(f"CWE-532 is detected in method, {debugLogger.fullName}") 
```

### Quark Script Result

```
$ python CWE-532.py 
CWE-532 is detected in method, Lcom/google/firebase/auth/FirebaseAuth; d (Lc/c/b/h/o;)V
```

### CODE REVIEW

**Comment by 18z** [comment]
hi @pulorsok 

I suggest rewriting the script as below.
Better naming, better structure, and is easier to understand.

```python
from quark.script import findMethodInAPK

SAMPLE_PATH = "dvba.apk"
TARGET_METHOD = [
    "Landroid/util/Log;",                       # class name
    "d",                                        # method name
    "(Ljava/lang/String; Ljava/lang/String;)I"  # descriptor
]
CREDENTIAL_KEYWORDS = [
    "token",
    "decrypt",
    "password"
]

methodsFound = findMethodInAPK(SAMPLE_PATH, TARGET_METHOD)

for debugLogger in methodsFound:
    arguments = debugLogger.getArguments(TARGET_METHOD)

    for keyword in CREDENTIAL_KEYWORDS:
        if keyword in arguments[1]:
            print(f"CWE-532 is detected in method, {debugLogger.fullName}") 
```

**Comment by 18z** [comment]
> We will keep updating the detect keywords list for better accuracy in detecting sensitive data.

No, we won't.
We encourage users to maintain their keyword list.

**Comment by 18z** [comment]
One thing I found strange

As shown in our documentation

```python
methodInstance.getArguments(none)
```

We don't need ```TARGET_METHOD``` for using getArguments() because you already have the method instance.

So I suggest a little change from my previous suggestion:

```python
from quark.script import findMethodInAPK

SAMPLE_PATH = "dvba.apk"
TARGET_METHOD = [
    "Landroid/util/Log;",                       # class name
    "d",                                        # method name
    "(Ljava/lang/String; Ljava/lang/String;)I"  # descriptor
]
CREDENTIAL_KEYWORDS = [
    "token",
    "decrypt",
    "password"
]

methodsFound = findMethodInAPK(SAMPLE_PATH, TARGET_METHOD)

for debugLogger in methodsFound:
    arguments = debugLogger.getArguments()

    for keyword in CREDENTIAL_KEYWORDS:
        if keyword in arguments[1]:
            print(f"CWE-532 is detected in method, {debugLogger.fullName}") 
```

**Comment by pulorsok** [comment]
Got it! I will update the changes later.

**Comment by pulorsok** [comment]
Hi @18z 
I've already modified the document and API spec. 
Please take a look. Thank you.

**Comment by sidra-asa** [comment]
@pulorsok 

Can you commit tests with findMethodInAPK function ?
And if possible, add a test case with [new-added code in quark script](https://github.com/pulorsok/quark-engine/blame/b70dd50770e2f7f6a1d5a247f4bacf61767325da/quark/script/__init__.py#L128-L166).

**Comment by 18z** [comment]
@pulorsok 

Please add authors who have contributed to this document.
https://github.com/quark-engine/quark-engine/blob/master/docs/source/conf.py#L22

And please also update the doc version to v22.10.2. (The second change in Quark Repo in October 2022)
https://github.com/quark-engine/quark-engine/blob/master/docs/source/conf.py#L25

**Comment by pulorsok** [comment]
@18z @sidra-asa 
I've already updated the document's version and author.
And added the test unit for new APIs. Please have a look. Thank you.

File: quark/script/__init__.py [hunk 1]
```diff
@@ -387,3 +426,44 @@ def getActivities(samplePath: PathLike) -> List[Activity]:
...
+
+def findMethodInAPK(
+    samplePath: PathLike,
+    targetMethod: Union[List[str], Method]
+) -> Method:
+    """Find the target method in APK.
+
+    :param samplePath: target file
+    :param targetMethod: python list contains class name,
+                         method name, and descriptor of target method
```
Comment (sidra-asa) [inline quark/script/__init__.py:439]: Typo

**Comment by haeter525** [comment]
Thank @pulorsok for this PR!

**Pull request merged**